File: src/parser/regexParser.cpp
#include "parser/regexParser.hpp"

using namespace Parser;

RegexParser::RegexParser(const std::regex* regex) { this->regex = regex; }

State* RegexParser::Run(State* state) {
    if (state->IsError) {
        return state;
    }

    if (state->Input.empty() || (unsigned int)state->Input.size() < (unsigned int)state->Index + 1) {
        return setErrorState(state, "RegexParser: got unexpected end of input");
    }

    std::smatch match;
    auto toCheck = state->Input.substr(state->Index);

    if (!std::regex_search(toCheck, match, *regex)) {
        return setErrorState(state, "RegexParser: couldn't match regex at index '" + std::to_string(state->Index) + "'");
    }

    if (match.empty()) {
        return setErrorState(state, "RegexParser: couldn't match regex at index '" + std::to_string(state->Index) + "'");
    }

    auto output = std::vector<std::string>();

    for (auto const& matchedString : match) {
        if (matchedString.matched) {
            output.push_back(matchedString.str());
        }
    }

    state->ResultMultiple = output;
    state->Index = state->Index + state->Result.size();
    state->IsError = false;
    state->Error = "";
    return state;
}

File: cmd/sealasm/parser/regex_parser.py
import re
from typing import Callable, Optional

from parser.base_parser import BaseParser
from parser.state import State


class RegexParser(BaseParser):
    "used to check a given string only contains hex-allowed values"

    def __init__(self, regex: str, map_method: Optional[Callable] = None):
        self._regex: Pattern[AnyStr] = re.compile(regex)
        super().__init__(map_method)

    def run(self, state: State) -> State:
        if state.is_error:
            return state

        if not len(state.source) or len(state.source) < state.index + 1:
            return self._set_error_state(state, "RegexParser: got unexpected end of input")

        source = state.source[state.index:]
        output = self._regex.match(source)

        if output:
            state.result = output.string[output.start():output.end()]
            state.index = state.index + output.end()
            state.is_error = False
            state.error = ""
        else:
            state.error = f"RegexParser: couldn't match digits at index '{state.index}'"
            state.is_error = True

        return State(state=state).map(self._map_method)
